onnxruntime/contrib_ops/webgpu/bert/group_query_attention.cc

```diff
@@ -32,81 +32,103 @@ ONNX_OPERATOR_KERNEL_EX(
+  const auto& query = sh.AddOutput("query", ShaderUsage::UseUniform);
+  const auto& key = sh.AddOutput("key", ShaderUsage::UseUniform);
+  const auto& value = sh.AddOutput("val", ShaderUsage::UseUniform);
+
+  sh.MainFunctionBody() << "  // Dispatch over Q domain only\n"
+                        << "  let q_flat_idx = global_idx;\n"
+                        << "  let batch_idx = q_flat_idx / (uniforms.sequence_length * uniforms.hidden_size);\n"
+                        << "  let remainder = q_flat_idx % (uniforms.sequence_length * uniforms.hidden_size);\n"
+                        << "  let seq_idx = remainder / uniforms.hidden_size;\n"
+                        << "  let hidden_idx = remainder % uniforms.hidden_size;\n"
+                        << "\n"
+                        << "  // Calculate base offset for this (batch, seq) position in packed_qkv\n"
+                        << "  // Layout per token: [Q(hidden_size), K(kv_hidden_size), V(kv_hidden_size)]\n"
+                        << "  let token_size = uniforms.hidden_size + 2u * uniforms.kv_hidden_size;\n"
+                        << "  let base_offset = batch_idx * uniforms.sequence_length * token_size + seq_idx * token_size;\n"
+                        << "\n"
+                        << "  // Read and write Q\n"
+                        << "  let q_offset = base_offset + hidden_idx;\n"
+                        << "  let q_data = " << packed_qkv.GetByOffset("q_offset") << ";\n"
+                        << "  " << query.SetByIndices("vec3<u32>(batch_idx, seq_idx, hidden_idx)", "q_data") << ";\n"
+                        << "\n"
+                        << "  // Process K and V if within kv_hidden_size range\n"
+                        << "  if (hidden_idx < uniforms.kv_hidden_size) {\n"
+                        << "    let k_offset = base_offset + uniforms.hidden_size + hidden_idx;\n"
+                        << "    let k_data = " << packed_qkv.GetByOffset("k_offset") << ";\n"
+                        << "    " << key.SetByIndices("vec3<u32>(batch_idx, seq_idx, hidden_idx)", "k_data") << ";\n"
+                        << "\n"
+                        << "    let v_offset = base_offset + uniforms.hidden_size + uniforms.kv_hidden_size + hidden_idx;\n"
+                        << "    let v_data = " << packed_qkv.GetByOffset("v_offset") << ";\n"
+                        << "    " << value.SetByIndices("vec3<u32>(batch_idx, seq_idx, hidden_idx)", "v_data") << ";\n"
+                        << "  }\n";
   return Status::OK();
 }
 
 Status SplitPackedQKV(onnxruntime::webgpu::ComputeContext& context, const WebgpuAttentionParameters& params, const Tensor* packedQKV, Tensor* query, Tensor* key, Tensor* val) {
   SplitPackedQKVProgram program;
-  auto input_size = packedQKV->Shape().Size();
+  // Dispatch only over Q domain: batch_size * sequence_length * hidden_size
+  auto q_size = static_cast<uint32_t>(params.batch_size_ * params.sequence_length_ * params.hidden_size_);
```

Comment:
If you use components when AddInput/AddOutputs, it may bring further perf improvement.